### PR TITLE
Refine insert mode to correctly capture all shadow DOM focuses/blurs

### LINKED ISSUE #2504: Use of shadowRoot in page makes each focus event register another focus handler, potentially building up to massive slowdowns
I believe this line is the culprit: https://github.com/philc/vimium/blob/master/content_scripts/mode_insert.coffee#L69

Repro:

1) Open http://openfin.github.io/fin-hypergrid-polymer-demo/components/fin-hypergrid/demo.html
2) Add a breakpoint in the `else if (event.target.shadowRoot) {` block of mode_insert.js
3) Click a cell, observe that the event handler gets invoked a couple times (not sure why it's already more than 1).  Click another cell, observe that it now gets invoked 3 times.  Click another, 4 times, etc etc.

Vimium 1.59

### PR DESCRIPTION
**Pull request by mrmr1993**
This replaces #2505 to (hopefully) fully fix #2504.

### CODE REVIEW

File: content_scripts/mode_insert.coffee [hunk 1]
```diff
@@ -47,34 +47,61 @@ class InsertMode extends Mode
...
-            _name: "shadow-DOM-input-mode"
-            blur: (event) ->
-              if event.target.shadowRoot == shadowRoot
-                handlerStack.remove()
-                for own type, listener of eventListeners
-                  shadowRoot.removeEventListener type, listener, true
+        # NOTE(mrmr1993): The first element of event.path gives us the element being focused, even when it is
+        # in an (open) shadow DOM.
+        target = event.path?[0] ? event.target
+        if ShadowRoot?
```
Comment (gdh1995) [inline content_scripts/mode_insert.coffee]: Here `window.ShadowRoot?` is better: Chrome has `window.ShadowRoot` as early as version 36, but Firefox only declares a global `ShadowRoot` if `dom.webcomponents.enabled` is `true` in `about:config`.
BTW, now Vimium can not work if no `ShadowDOM` APIs, so in face this line is not needed.

Comment (mrmr1993) [inline content_scripts/mode_insert.coffee]: > Here `window.ShadowRoot?` is better

Fair enough.

> Firefox only declares a global `ShadowRoot` if `dom.webcomponents.enabled` is `true` in `about:config`.

I know.

> BTW, now Vimium can not work if no `ShadowDOM` APIs, so in face this line is not needed.

It works fine in Firefox and in PhantomJS, both with no Shadow DOM.

Comment (gdh1995) [inline content_scripts/mode_insert.coffee]: Oh it's my mistake to think VImium can not work without ShadowDOM. So this line is indeed necessary.

**Comment by philc** [comment]
This no longer merges cleanly, but it doesn't look too hard to merge. I'm not sure if the underlying issue is still a problem. @gdh1995 what do you think?

**Comment by gdh1995** [comment]
Oh yes this has been replaced by #2851 , which is from the idea of https://github.com/philc/vimium/issues/2841#issuecomment-347995300

**Comment by philc** [comment]
Thanks for tying those all together @gdh1995. Closing this.

**Pull request closed without merging**